core/store/src/trie/mod.rs

```diff
@@ -495,16 +500,24 @@ impl Trie {
         Some(PartialStorage { nodes: PartialState(nodes) })
     }
 
-    pub fn from_recorded_storage(partial_storage: PartialStorage) -> Self {
+    pub fn from_recorded_storage(partial_storage: PartialStorage, root: StateRoot) -> Self {
         let recorded_storage =
             partial_storage.nodes.0.into_iter().map(|value| (hash(&value), value)).collect();
-        Trie {
-            storage: Box::new(TrieMemoryPartialStorage {
-                recorded_storage,
-                visited_nodes: Default::default(),
-            }),
-            flat_state: None,
-        }
+        let storage = Box::new(TrieMemoryPartialStorage {
+            recorded_storage,
+            visited_nodes: Default::default(),
+        });
+        Self::new(storage, root, None)
+    }
+
+    pub fn get_root(&self) -> &StateRoot {
+        &self.root
+    }
+
+    /// Sets state root of current Trie to given value.  Used only in tests.
+    #[doc(hidden)]
+    pub fn set_root(&mut self, root: StateRoot) {
```

Comment:
I think we shouldn't be able to do this even in tests. When we switch to flat state, after setting new root flat state may become incorrect, and we may not be able to create new "instance".

I suggest to create new `Trie` for each separate root, as it is handled in the code in the same way and it is cheap. Thinking out loud - we could emphasize it by renaming `Trie` to smth like `TrieSnapshot` to say that this struct doesn't correspond to the whole persistent trie.

Relevant discussion https://github.com/near/nearcore/pull/7345#discussion_r938933995

Reply:
Currently FlatState::get_ref takes root as argument (even though it ignores it at the moment?) so my understanding is that this should be safe to change the root?

Reply:
Oh, it shouldn't take it. Feel free to remove it, or I can do it in the following PR